### PR TITLE
Assorted touch fixes

### PR DESCRIPTION
**Pull request by yol**
I discovered issues with the touch subsystem while doing Wayland implementation tests (pkerling/xbmc#43), some of which are fixed by this PR

## Description
* Inertial scroll
  * Math does not add up (70b930fa) resulting in abrupt stop
  * Scrolling is not actually "inertial", i.e. if user starts touch, scrolls around a bit, then holds the finger for one second and releases, inertial scrolling will still kick in despite there not being any movement - very annoying (1d1c69f)
* Gesture detection
  * When windowing does not provide size, detection is way too sensitive and often turns taps into pan gestures (843e2b41)
  * Multi-pointer taps miss the check for whether the pointer moved even when size is provided (3d83da4) and so also frequently get turned into gestures
* Gestures are not aborted on window switch, resulting in unusable settings windows (because list control does not release exclusive mouse mode) when two-finger swipe gesture is used to go back to the previous window (f51f7bbd)

Rest of the commits are cleanup and C++11.

## Motivation and Context
Improve touch experience

## How Has This Been Tested?
Run-tested on Linux/Wayland

Android testing would be much appreciated. I can do it if someone tells me how I can get the resulting apk from jenkins?

## Types of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by MartijnKaijser** [comment]
http://jenkins.kodi.tv/job/BuildMulti-All/821/  builds will end up in test-builds

File: xbmc/input/touch/generic/GenericTouchActionHandler.cpp [hunk 1]
```diff
@@ -179,8 +182,8 @@ void CGenericTouchActionHandler::focusControl(float x, float y)
 {
   XBMC_Event newEvent{XBMC_SETFOCUS};
 
-  newEvent.focus.x = (uint16_t)x;
-  newEvent.focus.y = (uint16_t)y;
+  newEvent.focus.x = static_cast<int> (std::round(x));
```
Comment (notspiff) [inline xbmc/input/touch/generic/GenericTouchActionHandler.cpp]: This is not equivalent.

Comment (pkerling) [inline xbmc/input/touch/generic/GenericTouchActionHandler.cpp]: Yes. I think that it should have been rounded in the first place.

File: xbmc/input/touch/generic/GenericTouchInputHandler.cpp [hunk 2]
```diff
@@ -28,7 +28,13 @@
 #include "threads/SingleLock.h"
 #include "utils/log.h"
 
-#define TOUCH_HOLD_TIMEOUT  1000
+namespace
+{
+
```
Comment (Montellese) [inline xbmc/input/touch/generic/GenericTouchInputHandler.cpp]: Please remove the additional blank lines.

**Review by Montellese — APPROVED** [review]
Looks good apart from the minor. I can't comment on the inertial scrolling, that's @Memphiz' baby.

**Comment by Memphiz** [comment]
Tested on ios - seems to work fine still. I am fine with removing the decrease of deacceleration - feels good enough without imo. I added it in the first place to mimic what apple does as good as possible (the 0.9 decrease factor was determined by stomach feeling ...

**Review by Memphiz — APPROVED** [review]

File: xbmc/input/linux/LinuxInputDevices.h [hunk 3]
```diff
@@ -87,9 +86,9 @@ class CLinuxInputDevice
   bool m_bUnplugged;
   std::deque<XBMC_Event> m_equeue;
   int m_mt_currentSlot;
-  int m_mt_x[TOUCH_MAX_POINTERS];
-  int m_mt_y[TOUCH_MAX_POINTERS];
-  TouchInput m_mt_event[TOUCH_MAX_POINTERS];
+  int m_mt_x[CGenericTouchInputHandler::MAX_POINTERS];
```
Comment (Paxxi) [inline xbmc/input/linux/LinuxInputDevices.h]: If we're modernizing it would make sense to make these `std::aray` but might be out of scope for this PR.

File: xbmc/input/touch/generic/GenericTouchActionHandler.cpp [hunk 4]
```diff
@@ -161,43 +166,28 @@ int CGenericTouchActionHandler::QuerySupportedGestures(float x, float y)
...
 void CGenericTouchActionHandler::focusControl(float x, float y)
 {
-  XBMC_Event newEvent;
-  memset(&newEvent, 0, sizeof(newEvent));
+  XBMC_Event newEvent{XBMC_SETFOCUS};
 
-  newEvent.type = XBMC_SETFOCUS;
-  newEvent.focus.x = (uint16_t)x;
-  newEvent.focus.y = (uint16_t)y;
+  newEvent.focus.x = static_cast<int> (std::round(x));
```
Comment (Paxxi) [inline xbmc/input/touch/generic/GenericTouchActionHandler.cpp]: why is this changed from uint16_t to int?

File: xbmc/input/touch/generic/GenericTouchInputHandler.cpp [hunk 5]
```diff
@@ -361,44 +363,40 @@ void CGenericTouchInputHandler::triggerDetectors(TouchInput event, int32_t point
...
       break;
     }
 
     default:
       return;
   }
 
-  std::set<IGenericTouchGestureDetector*> finishedDetectors;
-  for (std::set<IGenericTouchGestureDetector*>::const_iterator detector = m_detectors.begin(); detector != m_detectors.end(); ++detector)
+  for (auto it = m_detectors.begin(); it != m_detectors.end(); )
```
Comment (Paxxi) [inline xbmc/input/touch/generic/GenericTouchInputHandler.cpp]: might as well switch this to `std::remove_if` when changing it.

**Review by Paxxi — COMMENTED** [review]
I trust @Montellese review on the functional changes, added a few nitpicks that are optional.

Can try and test the changes on Windows over the weekend.

File: xbmc/input/touch/generic/GenericTouchActionHandler.cpp (hunk 4 above)
Comment (pkerling) [inline xbmc/input/touch/generic/GenericTouchActionHandler.cpp]: XBMC_SetFocusEvent has `int`s, why cut it to 16 bit? https://github.com/xbmc/xbmc/blob/master/xbmc/windowing/XBMC_events.h#L106

File: xbmc/input/linux/LinuxInputDevices.h (hunk 3 above)
Comment (pkerling) [inline xbmc/input/linux/LinuxInputDevices.h]: I have to stop somewhere or this never ends ;-)

File: xbmc/input/touch/generic/GenericTouchInputHandler.cpp (hunk 5 above)
Comment (pkerling) [inline xbmc/input/touch/generic/GenericTouchInputHandler.cpp]: `remove_if` does not work with `std::set` because it is ordered and `remove_if` needs to reorder elements. I see no strong reason why `m_detectors` would need to be a `std::set` though. Not like you can try to push two identical `unique_ptr`s anyway. Should we make it a `std::vector`?

File: xbmc/input/touch/generic/GenericTouchInputHandler.cpp (hunk 2 above)
Comment (pkerling) [inline xbmc/input/touch/generic/GenericTouchInputHandler.cpp]: done

**Comment by MartijnKaijser** [comment]
It feels like it improves touch quite a lot. Hitting buttons is sometimes still hit and miss but haven't tested extensively yet. Selecting in listings is still quite difficult to get the correct item selected but that could be a skinning engine issue. The highlighted item while scrolling also seems a bit flaky.

**Comment by fritsch** [comment]
One thing at a time, please. If this one brings us forward, what it does,
squash together and we are good to go.

With the signed integer take care no one uses the center in the middle :-)

Am 05.08.2017 8:50 vorm. schrieb "Martijn Kaijser" <notifications@github.com
>:

It feels like it improves touch quite a lot. Hitting buttons is sometimes
still hit and miss but haven't tested extensively yet. Selecting in
listings is still quite difficult to get the correct item selected but that
could be a skinning engine issue. The highlighted item while scrolling also
seems a bit flaky.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/xbmc/xbmc/pull/12617#issuecomment-320419887>, or mute
the thread

**Comment by MartijnKaijser** [comment]
@fritsch i was pointing out that one of the issues is still kinda there. So that is still that "one thing at a time".

**Comment by pkerling** [comment]
I've identified potential problems with the gesture abort on window switch:

* I handled the gesture events in `CGUIWindowManager::OnApplicationMessage` - but the inertial scrolling handler might modify the events afterwards (drop some), so I think handling this in `CGUIWindowManager::OnAction` instead would be better. I'll add a fixup.
* What about other events that cause the focus and exclusive mouse control of a control to be broken no matter what it says - like dialogs (I get unresponsive dialogs from time to time, but haven't been able to pinpoint the problem)? Should we send gesture abort and inhibit likewise when `RegisterDialog` is called? Always or only under some circumstances, like only for modal dialogs or only when the topmost dialog changes?
  * Speaking of order: Some functions in window manager sort the dialogs based on their render order, like `RenderPass`. But most do not. `OnAction` for example doesn't do it. Won't that cause a mismatch between the topmost dialog on screen and the one that receives input?

**Comment by MartijnKaijser** [comment]
@Memphiz could you try again on iOS if all still works? Android has improved a lot with this. @Paxxi if you could try on Windows?

**Comment by Memphiz** [comment]
Is there a test build alteady?

**Comment by MartijnKaijser** [comment]
Building now

**Comment by pkerling** [comment]
squashed

**Comment by Memphiz** [comment]
Behaves fine on iOS imo.

**Comment by pkerling** [comment]
forgot to reorder the ARM32 fixed to before the std::round usage, so once again...
